src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,318 @@
+    lp :: Float64
+    "The evaluation function used to calculate the output."
+    f :: M
+end
+#############################
+# Various StatsBase methods #
+#############################
+
+function Base.show(io::IO, m::ModeResult)
+    println(io, Printf.@sprintf("ModeResult with minimized lp of %.2f", m.lp), "\n")
+    show(io, m.values)
+end
+
+function StatsBase.coeftable(m::ModeResult)
+    # Get columns for coeftable.
+    terms = StatsBase.coefnames(m)
+    estimates = m.values.array[:,1]
+    stderrors = StatsBase.stderror(m)
+    tstats = estimates ./ stderrors
+
+    StatsBase.CoefTable([estimates, stderrors, tstats], ["estimate", "stderror", "tstat"], terms)
+end
+
+function StatsBase.informationmatrix(m::ModeResult; hessian_function=ForwardDiff.hessian, kwargs...)
+    # Calculate Hessian and information matrix.
+    varnames = StatsBase.coefnames(m)
+    f = invlink(m.f)
+    info = inv(hessian_function(x -> f(x), m.values.array[:, 1]))
+    return NamedArrays.NamedArray(info, (varnames, varnames))
+end
+
+StatsBase.coef(m::ModeResult) = m.values
+StatsBase.coefnames(m::ModeResult) = names(m.values)[1]
+StatsBase.params(m::ModeResult) = StatsBase.coefnames(m)
+StatsBase.vcov(m::ModeResult) = StatsBase.informationmatrix(m)
+StatsBase.loglikelihood(m::ModeResult) = m.lp
+
+####################
+# Optim.jl methods #
+####################
+
+"""
+    Optim.optimize(model::Model, ::MLE, args...; kwargs...)
+
+Compute a maximum likelihood estimate of the `model`.
+
+# Examples
+
+```julia
+@model function f(x)
+    m ~ Normal(0, 1)
+    x ~ Normal(m, 1)
+end
+
+model = f(1.5)
+mle = optimize(model, MLE())
+
+# Use a different optimizer
+mle = optimize(model, MLE(), NelderMeade())
+```
+"""
+function Optim.optimize(model::Model, ::MLE, args...; kwargs...)
+    ctx = OptimizationContext(DynamicPPL.LikelihoodContext())
+    return optimize(model, OptimLogDensity(model, ctx, true), args...; kwargs...)
+end
+
+"""
+    Optim.optimize(model::Model, ::MAP, args...; kwargs...)
+
+Compute a maximum a posterior estimate of the `model`.
+
+# Examples
+
+```julia
+@model function f(x)
+    m ~ Normal(0, 1)
+    x ~ Normal(m, 1)
+end
+
+model = f(1.5)
+map_est = optimize(model, MAP())
+
+# Use a different optimizer
+map_est = optimize(model, MAP(), NelderMeade())
+```
+"""
+function Optim.optimize(model::Model, ::MAP, args...; kwargs...)
+    ctx = OptimizationContext(DynamicPPL.DefaultContext())
+    return optimize(model, OptimLogDensity(model, ctx, true), args...; kwargs...)
+end
+
+"""
+    Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.LBFGS(), args...; kwargs...)
+
+Estimate a mode, i.e., compute a MLE or MAP estimate.
+"""
+function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.LBFGS(), args...; kwargs...)
+    # Do some initialization.
+    spl = DynamicPPL.SampleFromPrior()
+    init_params = model(f.vi, spl)
+    init_vals = f.vi[spl]
+
+    # Optimize!
```

Comment:
Could we just use a closure here instead of the additional struct?

Reply:
I'm also wondering about this. Is there a reason why we don't use something like `make_logjoint` already used for VI: https://github.com/TuringLang/Turing.jl/blob/cb56cd2de7b9f398d5619d78369cd7ee3a2510fb/src/variational/VariationalInference.jl#L230-L257

And for MLE we could just wrap a `LikelihoodCtx`. 

With that being said, the current impl looks nice but might be overly complex with out much gain? :confused:

Reply:
With current Turing.jl you can do:
```julia
julia> @model function f(x)
           m ~ Normal(0, 1)
           x ~ Normal(m, 1)
       end
DynamicPPL.ModelGen{var"###generator#276",(:x,),(),Tuple{}}(##generator#276, NamedTuple())

julia> model = f(1.5)
DynamicPPL.Model{var"###evaluator#275",(:x,),Tuple{Float64},(),DynamicPPL.ModelGen{var"###generator#276",(:x,),(),Tuple{}}}(##evaluator#275, (x = 1.5,), DynamicPPL.ModelGen{var"###generator#276",(:x,),(),Tuple{}}(##generator#276, NamedTuple()))

julia> b = Turing.bijector(model);

julia> logπ = Turing.Variational.make_logjoint(model);

julia> objective = z -> - (logπ ∘ b)(z);

julia> map = optimize(objective, [0.], LBFGS());

julia> inv(b)(map.minimizer)
1-element Array{Float64,1}:
 0.7499999999941116

julia> map
 * Status: success

 * Candidate solution
    Minimizer: [7.50e-01]
    Minimum:   2.400377e+00

 * Found with
    Algorithm:     L-BFGS
    Initial Point: [1.00e+00]

 * Convergence measures
    |x - x'|               = 2.50e-01 ≰ 0.0e+00
    |x - x'|/|x'|          = 3.33e-01 ≰ 0.0e+00
    |f(x) - f(x')|         = 6.25e-02 ≰ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 2.60e-02 ≰ 0.0e+00
    |g(x)|                 = 3.67e-11 ≤ 1.0e-08

 * Work counters
    Seconds run:   0  (vs limit Inf)
    Iterations:    1
    f(x) calls:    3
    ∇f(x) calls:   3
```

You can also get the corresponding ranges by
```julia
julia> b, sym2val = Turing.bijector(model; sym_to_ranges = Val(true));

julia> sym2val
(m = UnitRange{Int64}[1:1],)
```
which could be used internally to construct the `NamedArray`.

Reply:
Could also use `Turing.gradient_logp` and pass this to `optimize` if there is some magic in there that we need.

Reply:
I suggested doing that and IMO we should change `make_logjoint` etc in the same way. It is much easier to hook into and analyze the internals if we don't use anonymous closures (see e.g. https://github.com/TuringLang/DynamicPPL.jl/issues/45).

Reply:
That makes sense, and I agree that the closure-idea isn't particularly nice (unless the user does it themselves, making things more transparent) :+1: 

Also, I guess this `OptimizationContext` will cover most use-cases + be slightly more efficient than a closure over `DefaultContext`.

Reply:
I'd like to note here that this is more or less what I originally had -- @devmotion requested a callable struct in the first place ([here](https://github.com/TuringLang/Turing.jl/pull/1230#discussion_r413531662)).

I have no preference either way but I don't much care for going back and forth.

Reply:
Currently I like @devmotion's original suggestion and would prefer to keep the callable struct. In fact, I think later on it should be generalized outside of the `Optim` context so that everyone can use it for whatever purposes they have.

Reply:
Make sense.